drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1612 @@
+       * the transceiver's response is complete. Length check ensures that
+       * we're not checking uninitialized memory in the case that `length` is
+       * 0.
+       */
+
+      if (length > 0 && buf[i - 1] == '\n')
+        {
+          line_complete = true;
+          break;
+        }
+    }
+
+  /* Insufficient buffer space to handle response */
+
+  if (!line_complete)
+    {
+      return -ENOBUFS;
+    }
+
+  /* Overwrite preceding \r with null terminator */
+
+  buf[i - 2] = '\0';
+  return i - 2; /* Number of bytes read excluding \r\n */
+}
+
+/****************************************************************************
+ * Name: mac_pause
+ *
+ * Description:
+ *   Pauses the MAC layer. Required before every transmission/receive.
+ *
+ ****************************************************************************/
+
+static int mac_pause(FAR struct rn2xx3_dev_s *priv)
+{
+  ssize_t length = 0;
+  char response[30];
+
+  /* Issue pause command */
+
+  length =
+      file_write(&priv->uart, "mac pause\r\n", sizeof("mac pause\r\n") - 1);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Wait for response of watchdog timer timeout */
+
+  length = read_line(priv, response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Check for pause duration */
+
+  if (strstr(response, MAC_PAUSE_DUR) == NULL)
+    {
+      return -EIO;
+    }
+
+  return 0;
+};
+
+/****************************************************************************
+ * Name: get_command_err
+ *
+ * Description:
+ *   Parses the command error response from the radio module and converts it
+ *   into an error code. 0 for 'ok', -EINVAL for 'invalid_param' and -EBUSY
+ *   for 'busy'. If an unknown error occurs, -EIO is returned.
+ *
+ ****************************************************************************/
+
+static int get_command_err(FAR struct rn2xx3_dev_s *priv)
+{
+  char response[18];
+  ssize_t length;
+
+  length = read_line(priv, response, sizeof(response));
+  if (length < 0)
+    {
+      return length;
+    }
+
+  if (strstr(response, "ok"))
+    {
+      /* Do nothing, this is good */
+
+      return 0;
+    }
+  else if (strstr(response, "invalid_param"))
+    {
+      wlerr("RN2xx3 invalid_param\n");
+      return -EINVAL;
+    }
+  else if (strstr(response, "busy"))
+    {
+      wlerr("RN2xx3 busy");
+      return -EBUSY;
+    }
+
+  wlerr("Unknown error");
+  return -EIO;
+}
+
+/****************************************************************************
+ * Name: rn2903_txpwrlevel
+ *
+ * Description:
+ *    Get the transmission power level that is greater than or equal to the
+ *    requested transmission power in dBm. These levels are for the RN2903.
+ *    The true dBm level that was set is returned in the `dbm` pointer.
+ *
+ ****************************************************************************/
+
+static int8_t rn2903_txpwrlevel(FAR float *dbm)
```

Comment:
are floats really required here? it's not friendly for MCU without FPU. You should avoid floats in shared code and use something more universal if possible, like `fixedmath.h`

Reply:
I feel like this is more intuitive than fixed math. Is it really that unfriendly for MCUs without an FPU? It is just used in a comparison.